Index created on 2.x causing "unsupported_operation_exception" in 5.x when used for sorting

<!-- Bug report -->

**Elasticsearch version** (`bin/elasticsearch --version`): 5.5.0

**Plugins installed**: []

**JVM version** (`java -version`): Java 8 u131

**OS version** (`uname -a` if on a Unix-like system): Ubuntu 16.04

**Description of the problem including expected versus actual behavior**:

**Steps to reproduce**:

## Create index in 2.x (I used 2.3.4)

```
DELETE test1
DELETE test2

PUT test1
{
  "mappings": {
    "_default_": {
      "properties": {
        "testfield": {
          "type": "string", "index": "not_analyzed"
        }
      }
    }
  }
}

PUT test2
{
  "mappings": {
    "_default_": {
      "properties": {
        "testfield": {
          "type": "string", "index": "not_analyzed"
        }
      }
    }
  }
}

POST test1/test
{
  "testfield": "my test"
}

# Don't index anything into test2

POST test*/_search
{
  "query": {
    "match_all": {}
  },
  "sort": [
    {
      "testfield": {
        "order": "asc",
        "unmapped_type": "string"
      }
    }
  ]
}
```

## Upgrade to 5.x (I used 5.5.0)

The query above, now results in: (Also if I change to "unmapped_type": "keyword")

```
{
  "took": 19,
  "timed_out": false,
  "_shards": {
    "total": 10,
    "successful": 5,
    "failed": 5,
    "failures": [
      {
        "shard": 0,
        "index": "test2",
        "node": "JBHJ1HPYQbKMOxJ1hSotrw",
        "reason": {
          "type": "unsupported_operation_exception",
          "reason": null
        }
      }
    ]
  },
  "hits": {
    "total": 1,
    "max_score": null,
    "hits": [
      {
        "_index": "test1",
        "_type": "test",
        "_id": "AV3Qh0G3rdu9hzgAfZn2",
        "_score": null,
        "_source": {
          "testfield": "my test"
        },
        "sort": [
          "my test"
        ]
      }
    ]
  }
}
```
and is producing this stack trace:

```
[2017-08-11T11:13:16,153][DEBUG][o.e.a.s.TransportSearchAction] [JBHJ1HP] [test2][0], node[JBHJ1HPYQbKMOxJ1hSotrw], [P], s[STARTED], a[id=OCMZA_bHTA-cdS6nEHIBgA]: Failed to execute [SearchRequest{searchType=QUERY_THEN_FETCH, indices=[test*], indicesOptions=IndicesOptions[id=38, ignore_unavailable=false, allow_no_indices=true, expand_wildcards_open=true, expand_wildcards_closed=false, allow_alisases_to_multiple_indices=true, forbid_closed_indices=true], types=[], routing='null', preference='null', requestCache=null, scroll=null, source={
  "query" : {
    "match_all" : {
      "boost" : 1.0
    }
  },
  "sort" : [
    {
      "testfield" : {
        "order" : "asc",
        "unmapped_type" : "keyword"
      }
    }
  ]
}}] lastShard [true]
org.elasticsearch.transport.RemoteTransportException: [JBHJ1HP][127.0.0.1:9300][indices:data/read/search[phase/query]]
Caused by: java.lang.UnsupportedOperationException
	at java.util.AbstractMap.put(AbstractMap.java:209) ~[?:1.8.0_131]
	at org.elasticsearch.index.mapper.KeywordFieldMapper$TypeParser.parse(KeywordFieldMapper.java:152) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.index.mapper.MapperService.unmappedFieldType(MapperService.java:727) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.FieldSortBuilder.build(FieldSortBuilder.java:260) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.SortBuilder.buildSort(SortBuilder.java:156) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.parseSource(SearchService.java:630) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createContext(SearchService.java:481) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:457) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:253) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:330) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:327) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.TransportService$7.doRun(TransportService.java:644) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:638) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-5.5.0.jar:5.5.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_131]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]
[2017-08-11T11:13:16,155][DEBUG][o.e.a.s.TransportSearchAction] [JBHJ1HP] [test2][2], node[JBHJ1HPYQbKMOxJ1hSotrw], [P], s[STARTED], a[id=kkIHhfYIRf6vXlBlebIpgg]: Failed to execute [SearchRequest{searchType=QUERY_THEN_FETCH, indices=[test*], indicesOptions=IndicesOptions[id=38, ignore_unavailable=false, allow_no_indices=true, expand_wildcards_open=true, expand_wildcards_closed=false, allow_alisases_to_multiple_indices=true, forbid_closed_indices=true], types=[], routing='null', preference='null', requestCache=null, scroll=null, source={
  "query" : {
    "match_all" : {
      "boost" : 1.0
    }
  },
  "sort" : [
    {
      "testfield" : {
        "order" : "asc",
        "unmapped_type" : "keyword"
      }
    }
  ]
}}] lastShard [true]
org.elasticsearch.transport.RemoteTransportException: [JBHJ1HP][127.0.0.1:9300][indices:data/read/search[phase/query]]
Caused by: java.lang.UnsupportedOperationException
	at java.util.AbstractMap.put(AbstractMap.java:209) ~[?:1.8.0_131]
	at org.elasticsearch.index.mapper.KeywordFieldMapper$TypeParser.parse(KeywordFieldMapper.java:152) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.index.mapper.MapperService.unmappedFieldType(MapperService.java:727) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.FieldSortBuilder.build(FieldSortBuilder.java:260) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.SortBuilder.buildSort(SortBuilder.java:156) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.parseSource(SearchService.java:630) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createContext(SearchService.java:481) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:457) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:253) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:330) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:327) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.TransportService$7.doRun(TransportService.java:644) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:638) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-5.5.0.jar:5.5.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_131]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]
[2017-08-11T11:13:16,157][DEBUG][o.e.a.s.TransportSearchAction] [JBHJ1HP] [test2][3], node[JBHJ1HPYQbKMOxJ1hSotrw], [P], s[STARTED], a[id=GyLowxvgSeO8ZabUK13b4g]: Failed to execute [SearchRequest{searchType=QUERY_THEN_FETCH, indices=[test*], indicesOptions=IndicesOptions[id=38, ignore_unavailable=false, allow_no_indices=true, expand_wildcards_open=true, expand_wildcards_closed=false, allow_alisases_to_multiple_indices=true, forbid_closed_indices=true], types=[], routing='null', preference='null', requestCache=null, scroll=null, source={
  "query" : {
    "match_all" : {
      "boost" : 1.0
    }
  },
  "sort" : [
    {
      "testfield" : {
        "order" : "asc",
        "unmapped_type" : "keyword"
      }
    }
  ]
}}] lastShard [true]
org.elasticsearch.transport.RemoteTransportException: [JBHJ1HP][127.0.0.1:9300][indices:data/read/search[phase/query]]
Caused by: java.lang.UnsupportedOperationException
	at java.util.AbstractMap.put(AbstractMap.java:209) ~[?:1.8.0_131]
	at org.elasticsearch.index.mapper.KeywordFieldMapper$TypeParser.parse(KeywordFieldMapper.java:152) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.index.mapper.MapperService.unmappedFieldType(MapperService.java:727) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.FieldSortBuilder.build(FieldSortBuilder.java:260) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.SortBuilder.buildSort(SortBuilder.java:156) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.parseSource(SearchService.java:630) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createContext(SearchService.java:481) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:457) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:253) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:330) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:327) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.TransportService$7.doRun(TransportService.java:644) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:638) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-5.5.0.jar:5.5.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_131]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]
[2017-08-11T11:13:16,161][DEBUG][o.e.a.s.TransportSearchAction] [JBHJ1HP] [test2][1], node[JBHJ1HPYQbKMOxJ1hSotrw], [P], s[STARTED], a[id=WEeR_WDhSW-Uxkys3ChfQg]: Failed to execute [SearchRequest{searchType=QUERY_THEN_FETCH, indices=[test*], indicesOptions=IndicesOptions[id=38, ignore_unavailable=false, allow_no_indices=true, expand_wildcards_open=true, expand_wildcards_closed=false, allow_alisases_to_multiple_indices=true, forbid_closed_indices=true], types=[], routing='null', preference='null', requestCache=null, scroll=null, source={
  "query" : {
    "match_all" : {
      "boost" : 1.0
    }
  },
  "sort" : [
    {
      "testfield" : {
        "order" : "asc",
        "unmapped_type" : "keyword"
      }
    }
  ]
}}] lastShard [true]
org.elasticsearch.transport.RemoteTransportException: [JBHJ1HP][127.0.0.1:9300][indices:data/read/search[phase/query]]
Caused by: java.lang.UnsupportedOperationException
	at java.util.AbstractMap.put(AbstractMap.java:209) ~[?:1.8.0_131]
	at org.elasticsearch.index.mapper.KeywordFieldMapper$TypeParser.parse(KeywordFieldMapper.java:152) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.index.mapper.MapperService.unmappedFieldType(MapperService.java:727) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.FieldSortBuilder.build(FieldSortBuilder.java:260) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.SortBuilder.buildSort(SortBuilder.java:156) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.parseSource(SearchService.java:630) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createContext(SearchService.java:481) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:457) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:253) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:330) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:327) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.TransportService$7.doRun(TransportService.java:644) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:638) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-5.5.0.jar:5.5.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_131]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]
[2017-08-11T11:13:16,167][DEBUG][o.e.a.s.TransportSearchAction] [JBHJ1HP] [test2][4], node[JBHJ1HPYQbKMOxJ1hSotrw], [P], s[STARTED], a[id=ka-M_4BJStSmPorrvv2_Cw]: Failed to execute [SearchRequest{searchType=QUERY_THEN_FETCH, indices=[test*], indicesOptions=IndicesOptions[id=38, ignore_unavailable=false, allow_no_indices=true, expand_wildcards_open=true, expand_wildcards_closed=false, allow_alisases_to_multiple_indices=true, forbid_closed_indices=true], types=[], routing='null', preference='null', requestCache=null, scroll=null, source={
  "query" : {
    "match_all" : {
      "boost" : 1.0
    }
  },
  "sort" : [
    {
      "testfield" : {
        "order" : "asc",
        "unmapped_type" : "keyword"
      }
    }
  ]
}}]
org.elasticsearch.transport.RemoteTransportException: [JBHJ1HP][127.0.0.1:9300][indices:data/read/search[phase/query]]
Caused by: java.lang.UnsupportedOperationException
	at java.util.AbstractMap.put(AbstractMap.java:209) ~[?:1.8.0_131]
	at org.elasticsearch.index.mapper.KeywordFieldMapper$TypeParser.parse(KeywordFieldMapper.java:152) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.index.mapper.MapperService.unmappedFieldType(MapperService.java:727) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.FieldSortBuilder.build(FieldSortBuilder.java:260) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.sort.SortBuilder.buildSort(SortBuilder.java:156) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.parseSource(SearchService.java:630) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createContext(SearchService.java:481) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:457) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:253) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:330) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:327) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69) ~[elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.transport.TransportService$7.doRun(TransportService.java:644) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:638) [elasticsearch-5.5.0.jar:5.5.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-5.5.0.jar:5.5.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_131]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]

```

As soon as I index a doc in `test2`, the error goes away and the query is working:

```
POST test2/test
{
  "testfield": "my test"
}

POST test*/_search
{
  "query": {
    "match_all": {}
  },
  "sort": [
    {
      "testfield": {
        "order": "asc",
        "unmapped_type": "keyword"
      }
    }
  ]
}
``` 
now results in:

```
{
  "took": 8,
  "timed_out": false,
  "_shards": {
    "total": 10,
    "successful": 10,
    "failed": 0
  },
  "hits": {
    "total": 2,
    "max_score": null,
    "hits": [
      {
        "_index": "test1",
        "_type": "test",
        "_id": "AV3Qh0G3rdu9hzgAfZn2",
        "_score": null,
        "_source": {
          "testfield": "my test"
        },
        "sort": [
          "my test"
        ]
      },
      {
        "_index": "test2",
        "_type": "test",
        "_id": "AV3Qku9k6j_T2EVH4QtG",
        "_score": null,
        "_source": {
          "testfield": "my test"
        },
        "sort": [
          "my test"
        ]
      }
    ]
  }
}
```

--- a/core/src/main/java/org/elasticsearch/index/mapper/MapperService.java
+++ b/core/src/main/java/org/elasticsearch/index/mapper/MapperService.java
@@ -724,7 +724,7 @@ public class MapperService extends AbstractIndexComponent implements Closeable {
             if (typeParser == null) {
                 throw new IllegalArgumentException("No mapper found for type [" + type + "]");
             }
-            final Mapper.Builder<?, ?> builder = typeParser.parse("__anonymous_" + type, emptyMap(), parserContext);
+            final Mapper.Builder<?, ?> builder = typeParser.parse("__anonymous_" + type, new HashMap<>(), parserContext);
             final BuilderContext builderContext = new BuilderContext(indexSettings.getSettings(), new ContentPath(1));
             fieldType = ((FieldMapper)builder.build(builderContext)).fieldType();
 

--- a/core/src/test/java/org/elasticsearch/index/mapper/MapperServiceTests.java
+++ b/core/src/test/java/org/elasticsearch/index/mapper/MapperServiceTests.java
@@ -167,7 +167,16 @@ public class MapperServiceTests extends ESSingleNodeTestCase {
     }
 
     public void testUnmappedFieldType() {
-        Version version = VersionUtils.randomVersionBetween(random(), Version.V_2_0_0, Version.CURRENT);
+        assertUnmappedFieldType(Version.CURRENT);
+    }
+
+    public void testUnmappedFieldTypeBWC() {
+        // test BWC with indices created in 2.x
+        Version version = VersionUtils.randomVersionBetween(random(), Version.V_2_0_0, Version.V_2_4_6);
+        assertUnmappedFieldType(version);
+    }
+
+    private void assertUnmappedFieldType(Version version) {
         MapperService mapperService =
             createIndex("index", Settings.builder().put("index.version.created", version).build()).mapperService();
         assertThat(mapperService.unmappedFieldType("keyword"), instanceOf(KeywordFieldType.class));

--- a/core/src/test/java/org/elasticsearch/index/mapper/MapperServiceTests.java
+++ b/core/src/test/java/org/elasticsearch/index/mapper/MapperServiceTests.java
@@ -29,13 +29,16 @@ import org.elasticsearch.index.IndexService;
 import org.elasticsearch.index.mapper.KeywordFieldMapper.KeywordFieldType;
 import org.elasticsearch.index.mapper.MapperService.MergeReason;
 import org.elasticsearch.index.mapper.NumberFieldMapper.NumberFieldType;
+import org.elasticsearch.plugins.Plugin;
 import org.elasticsearch.test.ESSingleNodeTestCase;
+import org.elasticsearch.test.InternalSettingsPlugin;
 import org.elasticsearch.test.VersionUtils;
 import org.hamcrest.Matchers;
 
 import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.util.Arrays;
+import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
@@ -49,6 +52,11 @@ import static org.hamcrest.Matchers.instanceOf;
 import static org.hamcrest.Matchers.startsWith;
 
 public class MapperServiceTests extends ESSingleNodeTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> getPlugins() {
+        return Collections.singletonList(InternalSettingsPlugin.class);
+    }
 
     public void testTypeNameStartsWithIllegalDot() {
         String index = "test-index";
@@ -179,10 +187,21 @@ public class MapperServiceTests extends ESSingleNodeTestCase {
     private void assertUnmappedFieldType(Version version) {
         MapperService mapperService =
             createIndex("index", Settings.builder().put("index.version.created", version).build()).mapperService();
-        assertThat(mapperService.unmappedFieldType("keyword"), instanceOf(KeywordFieldType.class));
-        assertThat(mapperService.unmappedFieldType("long"), instanceOf(NumberFieldType.class));
-        // back compat
-        assertThat(mapperService.unmappedFieldType("string"), instanceOf(KeywordFieldType.class));
+        if (version.after(Version.V_2_4_6)) {
+            assertThat(mapperService.unmappedFieldType("keyword"), instanceOf(KeywordFieldType.class));
+        } else {
+            assertThat(mapperService.unmappedFieldType("keyword"), instanceOf(StringFieldType.class));
+        }
+        if (version.after(Version.V_2_4_6)) {
+            assertThat(mapperService.unmappedFieldType("long"), instanceOf(NumberFieldType.class));
+        } else {
+            assertThat(mapperService.unmappedFieldType("long"), instanceOf(LegacyLongFieldMapper.LongFieldType.class));
+        }
+        if (version.after(Version.V_2_4_6)) {
+            assertThat(mapperService.unmappedFieldType("string"), instanceOf(KeywordFieldType.class));
+        } else {
+            assertThat(mapperService.unmappedFieldType("string"), instanceOf(StringFieldType.class));
+        }
         assertWarnings("[unmapped_type:string] should be replaced with [unmapped_type:keyword]");
     }
 

--- a/core/src/test/java/org/elasticsearch/index/mapper/MapperServiceTests.java
+++ b/core/src/test/java/org/elasticsearch/index/mapper/MapperServiceTests.java
@@ -20,6 +20,7 @@
 package org.elasticsearch.index.mapper;
 
 import org.elasticsearch.ExceptionsHelper;
+import org.elasticsearch.Version;
 import org.elasticsearch.common.compress.CompressedXContent;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.xcontent.XContentFactory;
@@ -29,6 +30,7 @@ import org.elasticsearch.index.mapper.KeywordFieldMapper.KeywordFieldType;
 import org.elasticsearch.index.mapper.MapperService.MergeReason;
 import org.elasticsearch.index.mapper.NumberFieldMapper.NumberFieldType;
 import org.elasticsearch.test.ESSingleNodeTestCase;
+import org.elasticsearch.test.VersionUtils;
 import org.hamcrest.Matchers;
 
 import java.io.IOException;
@@ -165,7 +167,9 @@ public class MapperServiceTests extends ESSingleNodeTestCase {
     }
 
     public void testUnmappedFieldType() {
-        MapperService mapperService = createIndex("index").mapperService();
+        Version version = VersionUtils.randomVersionBetween(random(), Version.V_2_0_0, Version.CURRENT);
+        MapperService mapperService =
+            createIndex("index", Settings.builder().put("index.version.created", version).build()).mapperService();
         assertThat(mapperService.unmappedFieldType("keyword"), instanceOf(KeywordFieldType.class));
         assertThat(mapperService.unmappedFieldType("long"), instanceOf(NumberFieldType.class));
         // back compat

--- a/core/src/test/java/org/elasticsearch/search/sort/FieldSortIT.java
+++ b/core/src/test/java/org/elasticsearch/search/sort/FieldSortIT.java
@@ -860,7 +860,16 @@ public class FieldSortIT extends ESIntegTestCase {
     }
 
     public void testIgnoreUnmapped() throws Exception {
-        Version version = VersionUtils.randomVersionBetween(random(), Version.V_2_0_0, Version.CURRENT);
+        assertIgnoreUnmapped(Version.CURRENT);
+    }
+
+    public void testIgnoreUnmappedBWC() throws Exception{
+        // test BWC with indices created in 2.x
+        Version version = VersionUtils.randomVersionBetween(random(), Version.V_2_0_0, Version.V_2_4_6);
+        assertIgnoreUnmapped(version);
+    }
+
+    private void assertIgnoreUnmapped(Version version) throws IOException {
         prepareCreate("test")
             .setSettings(Settings.builder().put(indexSettings()).put("index.version.created", version).build()).get();
 

--- a/core/src/test/java/org/elasticsearch/search/sort/FieldSortIT.java
+++ b/core/src/test/java/org/elasticsearch/search/sort/FieldSortIT.java
@@ -23,12 +23,14 @@ import org.apache.lucene.util.BytesRef;
 import org.apache.lucene.util.LuceneTestCase;
 import org.apache.lucene.util.TestUtil;
 import org.apache.lucene.util.UnicodeUtil;
+import org.elasticsearch.Version;
 import org.elasticsearch.action.admin.indices.alias.Alias;
 import org.elasticsearch.action.index.IndexRequestBuilder;
 import org.elasticsearch.action.search.SearchPhaseExecutionException;
 import org.elasticsearch.action.search.SearchResponse;
 import org.elasticsearch.action.search.ShardSearchFailure;
 import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.common.xcontent.XContentFactory;
 import org.elasticsearch.common.xcontent.XContentType;
@@ -39,6 +41,7 @@ import org.elasticsearch.plugins.Plugin;
 import org.elasticsearch.search.SearchHit;
 import org.elasticsearch.test.ESIntegTestCase;
 import org.elasticsearch.test.InternalSettingsPlugin;
+import org.elasticsearch.test.VersionUtils;
 import org.hamcrest.Matchers;
 
 import java.io.IOException;
@@ -857,7 +860,9 @@ public class FieldSortIT extends ESIntegTestCase {
     }
 
     public void testIgnoreUnmapped() throws Exception {
-        createIndex("test");
+        Version version = VersionUtils.randomVersionBetween(random(), Version.V_2_0_0, Version.CURRENT);
+        prepareCreate("test")
+            .setSettings(Settings.builder().put(indexSettings()).put("index.version.created", version).build()).get();
 
         client().prepareIndex("test", "type1", "1").setSource(jsonBuilder().startObject()
                 .field("id", "1")

